### PR TITLE
SG-6302 better path handling

### PR DESCRIPTION
**Pull request by pscadding**
This addresses tickets:
SG-6302
SG-4463
SG-6301

Updates:

- Now syncs path cache before trying to find context folders.
- Now will try to get the folder path from the `PublishedFile` entity `path` field, and highlight the file in the OS's prefered file manager. (Except Linux see below).
- Now tries to open a step folder if one is found and is deeper than the context folder. Before if you tried to "Show in filesystem" for a task, and the schema only had `Step` folders then it would only take you as far as the parent entity folder.
- Updated to use `sgtk` instead of `tank`.

There are a few things in this pull request, that I'm not sure about.

1. When trying to figure out if there is a deeper step folder, I'm having to gather context paths for the Step, and then match them against the entity's context folders. Whilst this works in my tests, I'm not sure if there are any scenarios that might break this. For example, I'm assuming that the step can only have one folder registered under the parent context folder, maybe I should account for multiple? Also, I'm not sure when a context can have multiple paths returned from `context.filesystem_locations`.

2. When getting the path for the PublishedFile entities, I'm checking if the path is a valid folder on disk, if not then I'm checking if its a valid file on disk, failing that I try to account for published files that contain symbolic paths such as sequences, where they use %04d. As it stands the code checks to see if the parent dir is valid and then just open that. This is flawed as potentially the received path is neither symbolic or valid, but the parent folder is, in which case I will just open the parent dir either way. I wonder if I should make some attempt perhaps through the use of regex to understand if its a sequence path?

3. If we are opening a folder containing a file path, it would be nice to select the file in the window. I have the method for doing this on Mac and Windows, however, I couldn't find a reliable way for achieving this on Linux. The inconsistency concerns me, although I would prefer to highlight the file when possible rather than not doing it for windows or mac just because I couldn't in Linux. Perhaps someone knows of a reliable way to do it in Linux? Maybe we should separate the logic into a hook to allow clients to adjust it to their needs?

4. Similar to above but as I'm gathering sequences as folder paths rather than file paths, I can't select the files. Is this a problem?

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -60,17 +95,69 @@ def show_in_filesystem(self, entity_type, entity_ids):
         with the given entity ids.
         """
         paths = []
-        
+
+        self.sgtk.synchronize_filesystem_structure()
```
Comment (josh-t) [inline app.py:70]: suggest a comment on why we're doing this here.

File: app.py [hunk 2]
```diff
@@ -60,17 +95,69 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                # Find the publish so that we can get the path back.
+                published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", eid]], ['code', 'path'])
+                # Resolve the path for the local OS
+                try:
+                    local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+                except Exception as e:
+                    self.log_debug("Publish path couldn't be resolved falling back to context based path; reason: %s" % e)
+                    local_path = None
+
+                if local_path:
```
Comment (josh-t) [inline app.py]: could this all be done in one go? 

```python
if local_path and (os.path.isdir(local_path) or os.path.isfile(local_path)):
    paths.append(local_path)
```

File: app.py [hunk 3]
```diff
@@ -60,17 +95,69 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                        paths.append(local_path)
+                        continue
+
+                    # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
+                    # the parent folder of the path is valid and try opening that.
+                    # The issue with this logic is that possibly it was a directory that just didn't exist, as so
+                    # we would just be gathering the next directory up, ideally need a better way to handle sequences
+                    parent_dir = os.path.dirname(local_path)
+                    if parent_dir and os.path.isdir(parent_dir):
+                        paths.append(os.path.dirname(local_path))
```
Comment (josh-t) [inline app.py]: could just be `paths.append(parent_dir)` since you already have `parent_dir`

File: app.py [hunk 4]
```diff
@@ -60,17 +95,69 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+            if context.step:
+                # As steps are non project entities and not linked from the step to the other entities you won't get step
+                # paths back from the context.filesystem_locations
+                # If we have a step in the context, we should check to see if we can resolve the path more deeply using that.
+                step_paths = self.sgtk.paths_from_entity("Step", context.step['id'])
+
+                for context_path in context_paths:
+                    # Loop over the context paths, and check to see if any of them can be extended with a step path.
+                    # If it can't fall back to the standard context path
+                    step_path = next((step_path for step_path in step_paths if step_path.startswith(context_path)),
```
Comment (josh-t) [inline app.py]: I don't use `next` much so I'm not exactly sure that this is doing. Is this taking the first "step path" that starts with `context_path` or something else?

**Review by josh-t — APPROVED** [review]
Looks awesome. A few minor comments to consider.

File: app.py (hunk 4 above)
Comment (pscadding) [reply app.py]: That's exactly right, and I provide a default value of `None` so that if it doesn't find a match then it won't error. It would be similar to writing

```
step_path = None
for a_step_path in step_paths:
    if a_step_path.startswith(context_path):
        step_path = a_step_path
        break
```

Comment (pscadding) [reply app.py]: Which is related to my question number 1, which is to say I'm assuming we should only care about one match. I'm not sure on the situations where context_paths might be multiple, and if we would ever potentially get two step path matches.

File: app.py [hunk 5]
```diff
@@ -47,7 +53,36 @@ def launch(self, path):
...
         else:
             raise Exception("Platform '%s' is not supported." % system)
-        
+
+        self.log_debug("Executing command '%s'" % cmd)
+        exit_code = os.system(cmd)
+        if exit_code != 0:
+            self.log_error("Failed to launch '%s'!" % cmd)
+
+    def launch_file_folder(self, path):
```
Comment (manneohrstrom) [inline app.py]: please make this a protected method unless you intend to use it to the app's public API. In that case, it should be documented.

File: app.py [hunk 6]
```diff
@@ -47,7 +53,36 @@ def launch(self, path):
...
+        :param path: A file path
+        :return: None
+        """
+        self.log_debug("Launching file system viewer for file %s" % path)
+
+        # get the setting
+        system = sys.platform
+
+        # build a command that will open and ideally select the file in the OS's default file manager
+        if system == "linux2":
```
Comment (manneohrstrom) [inline app.py]: please update to `if system.startswith("linux")` everywhere.

File: app.py (hunk 5 above)
Comment (manneohrstrom) [inline app.py]: i find the naming of the method a little vague - the docstring talks about 'file manager' - perhaps `launch_filemanager_for_path()` or something is more easy to understand?

File: app.py [hunk 7]
```diff
@@ -47,7 +53,36 @@ def launch(self, path):
...
-        
+
+        self.log_debug("Executing command '%s'" % cmd)
+        exit_code = os.system(cmd)
+        if exit_code != 0:
+            self.log_error("Failed to launch '%s'!" % cmd)
+
+    def launch_file_folder(self, path):
+        """
+        This method will take a path to a file and open it in the OS's default file manager.
```
Comment (manneohrstrom) [inline app.py]: does this work for folder paths as well as paths to files? might be worth clarifying.

File: app.py [hunk 8]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
         for eid in entity_ids:
+
+            if entity_type == "PublishedFile":
+                # Find the publish so that we can get the path back.
+                published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", eid]], ['code', 'path'])
+                # Resolve the path for the local OS
+                try:
+                    local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+                except Exception as e:
+                    self.log_debug("Publish path couldn't be resolved falling back to context based path; reason: %s" % e)
```
Comment (manneohrstrom) [inline app.py]: what happens here? Should we push an error message back to the user? Please add a comment to clarify!

File: app.py [hunk 9]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                if local_path:
+                    if (os.path.isdir(local_path) or os.path.isfile(local_path)):
+                        # If the path is pointing to a valid file or folder then its good to use as is.
+                        paths.append(local_path)
+                        continue
+
+                    # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
+                    # the parent folder of the path is valid and try opening that.
+                    # The issue with this logic is that possibly it was a directory that just didn't exist, as so
+                    # we would just be gathering the next directory up, ideally need a better way to handle sequences
```
Comment (manneohrstrom) [inline app.py]: please add this as a `TODO` so we can grep for it!

File: app.py [hunk 10]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
                             
         if len(paths) == 0:
             self.log_info("No location exists on disk yet for any of the selected entities. "
                           "Please use shotgun to create folders and then try again!")
         else:
             # launch folder windows
             for x in paths:
-                self.launch(x)
-    
+                if os.path.isfile(str(x)):
```
Comment (manneohrstrom) [inline app.py]: when is `x` not a string?

File: app.py [hunk 11]
```diff
@@ -33,11 +33,17 @@ def init_app(self):
         self.engine.register_command("show_in_filesystem", self.show_in_filesystem, p)
             
     def launch(self, path):
-        self.log_debug("Launching file system viewer for folder %s" % path)        
-        
-        # get the setting        
+        """
+        This method will take a path to a folder and open it in the OS's default file manager.
```
Comment (manneohrstrom) [inline app.py]: this is not strictly true - this method just launches a path (file or folder) using it's associated application - in the case of `.ma` files, that would be maya, in the case of folders, that would be the OS file manager. Either please update the docstring or raise an exception if a path to a file is passed in!

File: app.py [hunk 12]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                # associated with this entity in the path cache.
+                paths.extend(context_paths)
                             
         if len(paths) == 0:
             self.log_info("No location exists on disk yet for any of the selected entities. "
                           "Please use shotgun to create folders and then try again!")
         else:
             # launch folder windows
             for x in paths:
-                self.launch(x)
```
Comment (manneohrstrom) [inline app.py:75]: It feels like this logic is in the wrong place, given the naming of these methods. I think i would prefer 

```
for x in paths:
   self.launch(x)
```

and then inside of `launch()` you can have one branch of logic handling folders and one handling files. That i think would end up being a lot cleaner.

File: app.py [hunk 13]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
         paths = []
-        
+
+        # As the this method relies on the path_cache to find the folders,
+        # we must check we are in sync before continuing
+        self.sgtk.synchronize_filesystem_structure()
+
         for eid in entity_ids:
+
+            if entity_type == "PublishedFile":
```
Comment (manneohrstrom) [inline app.py:76]: should there be a matching else here? What if `entity_type` is `Delivery` or `Version`? 
Also noticed there are several `continue` here, making the code hard to follow. Do you think you could avoid those by adding some `if..elif...else` - and ideally also generate a nice error message if the entity type isn't properly supported?

File: app.py [hunk 14]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                    self.log_debug("Publish path couldn't be resolved falling back to context based path; reason: %s" % e)
+                    local_path = None
+
+                if local_path:
+                    if (os.path.isdir(local_path) or os.path.isfile(local_path)):
+                        # If the path is pointing to a valid file or folder then its good to use as is.
+                        paths.append(local_path)
+                        continue
+
+                    # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
```
Comment (manneohrstrom) [inline app.py]: this logic feels like it's in the wrong place - shouldn't it be inside of `launch()`?

File: app.py [hunk 15]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+            context = self.sgtk.context_from_entity(entity_type, eid)
+            context_paths = context.filesystem_locations
+
+            if context.step:
+                # As steps are non project entities and not linked from the step to the other entities you won't get step
+                # paths back from the context.filesystem_locations
+                # If we have a step in the context, we should check to see if we can resolve the path more deeply using that.
+                step_paths = self.sgtk.paths_from_entity("Step", context.step['id'])
```
Comment (manneohrstrom) [inline app.py]: double quotes for strings pls. :)

Comment (manneohrstrom) [inline app.py]: This approach would result in 10000 matches on a TV episode with 10000 shots. I think we need this logic to be O(1) time complexity if possible.

File: app.py [hunk 16]
```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
...
+                        paths.append(parent_dir)
+                        continue
+
             # Use the path cache to look up all paths linked to the task's entity
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+            context = self.sgtk.context_from_entity(entity_type, eid)
+            context_paths = context.filesystem_locations
+
+            if context.step:
```
Comment (manneohrstrom) [inline app.py]: `context.filesystem_locations` is supposed to encapsulate the logic of 'which paths belong to my context' - it feels like we are trying to add extra logic here to work around the fact that it's not doing the right thing? Would it make sense to update core instead? Let's discuss.

**Review by manneohrstrom — CHANGES_REQUESTED** [review]
Did a quick review pass and have some questions and suggestions!

File: app.py (hunk 5 above)
Comment (pscadding) [reply app.py]: Yeah, I did wonder if it should be, I kept it open as the original `launch` method was public so I thought I would follow convention there, but happy to protect it.

File: app.py (hunk 6 above)
Comment (pscadding) [reply app.py]: Again this was copied from the launch method, I'll update both places.

File: app.py (hunk 8 above)
Comment (pscadding) [reply app.py]: I think this is a good point. My approach with adding the changes was to always fall back on the current behaviour (which is to use the context paths) if for whatever reason we could not find a better more specific way of showing the paths.
Perhaps your right, now it supports `PublishedFile` entities it should just fail if it can't find the path?

File: app.py (hunk 13 above)
Comment (pscadding) [reply app.py:76]: I'll take a look at the continues.

I did consider the `Version` and `Delivery` entities. 
The `Delivery` entity doesn't have an appropriate field for grabbing a path from, it instead links to `Versions` and `PublishedFiles`. The attachment field didn't seem to be suitable either, I couldn't add anything to mine, not sure what type of field it is, would need to check under the hood.
The `Version` entity does have an `Uploaded Movie`, and whilst this does support local file links, its usually used for uploaded movies which would not be compatible with this app, But I could add support for this field if we think it makes sense?

If we wanted to make this solution more complicated I guess we could allow entity/file field configuration, so the user could provide an entity type and a field code to check for files, but my preference would be to leave it for a feature request, to see if anyone wants that.

I'm not sure we should generate an error for other entity types, as they may have context folders, and so we should just fall back to the original behaviour of opening the folders generated from the schema.

File: app.py (hunk 11 above)
Comment (pscadding) [reply app.py]: That's fair enough, I guess its intended to only open folders, so I'll raise an exception. Is `ValueError` an appropriate error to raise, or should I be using a tank specific error?

File: app.py (hunk 10 above)
Comment (pscadding) [reply app.py]: I hit an issue with it not being a `str` at one point and needed to cast it to str, I should have commented why, as I can't replicate it now. I've removed the conversion for now.

File: app.py (hunk 14 above)
Comment (pscadding) [reply app.py]: I agree with you, and I've refactored it, however by doing this, it will now perform this hacky sequence check on all paths, not just ones that come from the PublishedFile entities.

File: app.py [hunk 17]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        Given a path this method will choose the appropriate way to open that as a folder in the OS's default
+        file manager. Folders will be opened as is
+        :param path:
+        :return: None
+        """
+
+        if os.path.isfile(path):
+            self._launch_filemanager_for_file(path)
+        elif os.path.isdir(path):
+            self._launch_filemanager_for_file(path)
```
Comment (manneohrstrom) [inline app.py]: should this be `_launch_filemanager_for_folder`?

File: app.py (hunk 16 above)
Comment (pscadding) [reply app.py]: Yeah, this is a good point, but one I can't really answer because I don't know what the ramifications of doing that might be.
I understand why the behaviour is not there by default, its because the step entities are non-project entities and are therefore not linked to a specific entity, so the only way I could find of finding the correct step folder, was to find all folders registered with that step and then see if it fits within the paths generated from the context. As you commented this approach could potentially be slow when dealing with lots of shots or assets (I've not tested).

Obviously, I wouldn't want to move a non-performant action into the core code. And at the moment I can't think of a better way of doing it. I added this feature (to open step folders) knowing full well that it was potentially flawed, so I'm OK with redacting this from the PR if we need to find a better solution. It would fit outside of the scope of my original task, so I would prefer to get the other fixes/features out.

File: app.py (hunk 15 above)
Comment (pscadding) [reply app.py]: I've put a comment in bellow regarding moving it to the core code, but I can't think of a better way of associating a step folder with the entity/context folders. I'll have a think some more, perhaps there are some clues or solutions already to be found within our core code?

File: app.py (hunk 6 above)
Comment (pscadding) [reply app.py]: Is it just the Linux checks I should do this for, or should I do it for Windows and Mac?

File: app.py [hunk 18]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
-            
+
     def launch(self, path):
-        self.log_debug("Launching file system viewer for folder %s" % path)        
-        
-        # get the setting        
+        """
+        Given a path this method will choose the appropriate way to open that as a folder in the OS's default
+        file manager. Folders will be opened as is
+        :param path:
```
Comment (manneohrstrom) [inline app.py]: missing description

File: app.py [hunk 19]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        :return: None
```
Comment (manneohrstrom) [inline app.py]: if it returns none, just omit the`:returns:` bit

File: app.py (hunk 18 above)
Comment (manneohrstrom) [reply app.py]: we should add a description saying that nothing will happen if the path is invalid.

File: app.py [hunk 20]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        elif os.path.isdir(path):
+            self._launch_filemanager_for_folder(path)
+        else:
+            # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
+            # the parent folder of the path is valid and try opening that.
+            # todo: The issue with this logic is that possibly it was a directory that just didn't exist,
+            # so we would just be gathering the next directory up,
+            # ideally need a better way to handle sequences.
+            parent_dir = os.path.dirname(path)
+            if parent_dir and os.path.isdir(parent_dir):
```
Comment (manneohrstrom) [inline app.py]: when would parent dir not be a directory here? do we need the isdir check? maybe worth adding a comment

File: app.py [hunk 21]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+                self._launch_filemanager_for_folder(path)
+
+            
+    def _launch_filemanager_for_folder(self, path):
+        """
+        This method will take a path to a folder and open it in the OS's default file manager.
+
+        :param path: A folder path
+        :return: None
```
Comment (manneohrstrom) [inline app.py]: can remove this line

File: app.py [hunk 22]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        """
+        self.log_debug("Launching file system viewer for folder %s" % path)
+
+        # Check that we don't have a file path.
+        if os.path.isfile(path):
+            raise ValueError(
+                "The path should be a folder not be a file path. "
+                "If you want to open a folder containing a file use _launch_filemanager_for_file. Path: %s" % path)
```
Comment (manneohrstrom) [inline app.py]: we should not include private methods in the error message. Just say 'path 'xxxx' is not a valid directory' is sufficient. I think the logic should be `if not os.path.isdir(path):` to ensure we check other things than just files too. Does that make sense?

File: app.py [hunk 23]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
             self.log_error("Failed to launch '%s'!" % cmd)
 
-    
+    def _launch_filemanager_for_file(self, path):
+        """
+        This method will take a path to a file and open it in the OS's default file manager.
+        This is only used for files, for folders use _launch_filemanager_for_folder.
+
+        :param path: A file path
+        :return: None
```
Comment (manneohrstrom) [inline app.py]: can remove this line.

File: app.py [hunk 24]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+            parent_dir = os.path.dirname(path)
+            if parent_dir and os.path.isdir(parent_dir):
+                self._launch_filemanager_for_folder(path)
+
+            
+    def _launch_filemanager_for_folder(self, path):
+        """
+        This method will take a path to a folder and open it in the OS's default file manager.
+
+        :param path: A folder path
```
Comment (manneohrstrom) [inline app.py]: add `:raises:`

File: app.py (hunk 23 above)
Comment (manneohrstrom) [reply app.py]: add `:raises:`

File: app.py [hunk 25]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+
+        :param path: A file path
+        :return: None
+        """
+        self.log_debug("Launching file system viewer for file %s" % path)
+
+        if not os.path.isfile(path):
+            raise ValueError(
+                "The path should be a file path not be a folder path. "
+                "If you want to open a folder use _launch_filemanager_for_folder. Path: %s" % path)
```
Comment (manneohrstrom) [inline app.py]: same feedback here as above

File: app.py [hunk 26]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        # get the setting
+        system = sys.platform
+
+        # build a command that will open and ideally select the file in the OS's default file manager.
+        if system.startswith("linux"):
+            # Can't find a reliable way to open a file browser and select a file on linux,
+            # so if we get passed a file path, only open up the folder.
+            cmd = 'xdg-open "%s"' % os.path.dirname(path)
+        elif system == "darwin":
+            cmd = 'open -R "%s"' % path
```
Comment (manneohrstrom) [inline app.py]: please add a comment explaining the -R option

File: app.py [hunk 27]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+        :param entity_id: int for the PublishedFile ID.
+        :return: str path or None depending on whether it successfully resolves the path.
+        """
+
+        # Find the publish so that we can get the path back.
+        published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", entity_id]], ["code", "path"])
+        # Resolve the path for the local OS
+        try:
+            local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+        except Exception as e:
```
Comment (manneohrstrom) [inline app.py]: can we use a less broad exception trap?

File: app.py [hunk 28]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+
+    def _get_published_file_path(self, entity_id):
+        """
+        Will attempt to return back a resolved path for a PublishedFile entity
+        :param entity_id: int for the PublishedFile ID.
+        :return: str path or None depending on whether it successfully resolves the path.
+        """
+
+        # Find the publish so that we can get the path back.
+        published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", entity_id]], ["code", "path"])
```
Comment (manneohrstrom) [inline app.py:47]: nit: code seems unused.

File: app.py [hunk 29]
```diff
@@ -31,46 +30,176 @@ def init_app(self):
...
+                # Use the path cache to look up all paths linked to the task's entity
+                context = self.sgtk.context_from_entity(entity_type, eid)
+                context_paths = context.filesystem_locations
+
+                if context.step:
+                    # As steps are non project entities and not linked from the step to the other entities
+                    # you won't get step paths back from the context.filesystem_locations.
+                    # If we have a step in the context,
+                    # we should check to see if we can resolve the path more deeply using that.
+                    step_paths = self.sgtk.paths_from_entity("Step", context.step["id"])
```
Comment (manneohrstrom) [inline app.py]: this will potentially return thousands of paths, it doesn't scale.

Could we instead say something like: if we are looking at a step, use shotgun to find all the tasks matching that entity and that step and then get the paths for those tasks. That should result in a constant number of matches even as a production scales up to hundreds of thousands of tasks.

**Review by manneohrstrom — CHANGES_REQUESTED** [review]
Looks great! We need to change the logic for getting the tasks paths though, i am worried that the current approach doesn't scale terribly well. See my comments - do you think that approach could work?

File: app.py (hunk 20 above)
Comment (pscadding) [reply app.py]: I think your right we do not want the `isdir` check. `dirname`, is just a string operation as far as I can tell and doesn't actually check for the existence of the directory, so I thought I should check for its actual existence, however looking back over this, its probably better to let the `_launch_filemanager_for_folder` raise the error if it failed to open the directory.

File: app.py (hunk 28 above)
Comment (pscadding) [reply app.py:47]: It's not clear in the GitHub review which line you are referring to, but I can't see any code that is not in use. Can you clarify which line? Cheers

File: app.py (hunk 27 above)
Comment (pscadding) [reply app.py]: I was treating the resolving of the publish path as extra functionality on top of the base context path resolving. So I was falling back to the context-based approach if for any reason the path couldn't be resolved. I agree I don't usually like using a blanket catch-all, I'll try and see if there is anything less broad.

File: app.py (hunk 29 above)
Comment (pscadding) [reply app.py]: Getting the `Task` using the `Step` and entity in the context shouldn't be a problem. However, you won't get any paths from the `Task` entities (unless you actually have `Task` folders in your schema). Our default is to just have step folders and not task ones. So I could check for tasks but it would only work if there were task folders as far as I can tell.

File: app.py (hunk 28 above)
Comment (manneohrstrom) [reply app.py:47]: sorry - the `code` field you are requesting from shotgun seems unused.

File: app.py (hunk 29 above)
Comment (manneohrstrom) [reply app.py]: We need to scale it down so that we don't get all the step paths. Can you get the entity first, and then based on that request the step folders somehow? So that we don't ask for *all* steps, since that doesn't scale.

File: app.py (hunk 28 above)
Comment (pscadding) [reply app.py:47]: Ah gotcha, heheh, I thought you meant some python code, not the code field its self. yep cool, I'll address that.

File: app.py (hunk 29 above)
Comment (pscadding) [reply app.py]: The path cache just doesn't contain anything that can link a step registry to the task it was associated with when the folder was created (understandably as steps can be linked to many tasks.) It also doesn't contain anything that links it to the entity either. I've not looked into how to do this yet, but perhaps I can generate the path for the step against the schema (the same way create folders does) and then maybe even check to see if the generated step folder exists in the path cache.

**Comment by pscadding** [comment]
I've removed the `Step` folder path ability, due to the performance concerns raised. I did some investigation into how it might be handled better but in the end, I didn't have time to find an acceptable solution.

Completing this Pull request should then mean that we add it to the env in the default config for published files. However, I noticed that the env serves both `PublishedFile` and `Version` entities. Currently, the code only adds support for `PublishedFile` entities. I had a look at `Versions` and there are two potential fields we could gather the path from: `Path to Frames` and `Path to Movie`. Both these fields are just string fields so they wouldn't support multi OS paths. I tested publishing a `Version` from Nuke using the publish2 and the submit for review, and it filled in both of these paths. However `Path to Movie` points to a mov file that doesn't exist within the review folder. So I assume this path will mostly be a dud. Should I add support for Versions even with these limitations? If I don't then without separating the `PublishedFiles` and Versions environments in the config, it will mean the app is shown as an AMI on Versions but won't show any folders at all.

File: app.py [hunk 30]
```diff
@@ -31,46 +30,158 @@ def init_app(self):
...
+            raise ValueError(
+                "The path \"%s\" is not a valid directory." % path)
+
+        # get the setting
         system = sys.platform
-        
+
         # run the app
-        if system == "linux2":
+        if system.startswith("linux2"):
```
Comment (jfboismenu) [inline app.py]: Do a startswith("linux"). There are some Linuxes where linux3 is returned.

Yeah, we should have a method in core that detects that...

File: app.py [hunk 31]
```diff
@@ -31,46 +30,158 @@ def init_app(self):
...
+            # -R causes the open command to open a finder window and select the file within the parent directory.
+            cmd = 'open -R "%s"' % path
+        elif system == "win32":
+            # /select makes windows select the file within the explorer window
+            cmd = 'explorer /select,"%s"' % path
+        else:
+            raise Exception("Platform '%s' is not supported." % system)
+
+        self.log_debug("Executing command '%s'" % cmd)
+        exit_code = os.system(cmd)
```
Comment (jfboismenu) [inline app.py]: I would feel much more comfortable if you insteade you wer passing in a array of arguments and using the subprocess module I can imagine how someone might be able to inject stuff inside your code by having " in the file name.

File: app.py (hunk 27 above)
Comment (jfboismenu) [reply app.py]: Maybe TankError?

File: app.py [hunk 32]
```diff
@@ -31,46 +30,158 @@ def init_app(self):
...
+        # As the this method relies on the path_cache to find the folders,
+        # we must check we are in sync before continuing.
+        self.sgtk.synchronize_filesystem_structure()
+
         for eid in entity_ids:
-            # Use the path cache to look up all paths linked to the task's entity
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+
+            file_path_found = False
```
Comment (jfboismenu) [inline app.py]: The following seems functionally equivalent and is easier to understand.

```python
pub_file_path = None
if entity_type == "PublishedFile":
    # If the entity type is a PublishedFile, we should try at first to extract the path of the publish
    # and open the folder for that. If we can't we will fall back on the context driven path.
    pub_file_path = self._get_published_file_path(eid)

if pub_file_path:
    paths.append(pub_file_path)
else:
    context = self.sgtk.context ...
```

**Review by jfboismenu — CHANGES_REQUESTED** [review]

**Comment by jfboismenu** [comment]
Sorry, I hit the button without providing general comments.

Looks good to me, but I agree with Manne's points and I have a few of my own. Almost there!

File: app.py (hunk 28 above)
Comment (pscadding) [reply app.py:47]: There is a bug with the `resolve_publish_path` which means that it requires the published_file to have the `code` key present, in certain situations. That bug is logged, so I'll leave this as it is a make a comment explaining, why the code is being gathered. @jfboismenu

File: app.py (hunk 31 above)
Comment (pscadding) [reply app.py]: Hey JF when you mean to pass an array of arguments and use the subprocess module, you mean instead of building a cmd string, pass each section of the cmd as a list of string to the subprocess module? 
The `_launch_filemanager_for_file` method its self would still be passed a string path though? So couldn't stuff still be injected into the file path name?

Comment (jfboismenu) [reply app.py]: Exactly. it would look something like `subprocess.call(["explorer", "/select,", path])`

If you really can't use subprocess, you need to consider using [pipes.quote](https://docs.python.org/2/library/pipes.html?highlight=quote#pipes.quote) which will escape the path properly for shell usage.

**Comment by pscadding** [comment]
OK changes submitted hopefully this OK?

File: app.py [hunk 33]
```diff
@@ -31,46 +32,168 @@ def init_app(self):
...
+        # get the setting
         system = sys.platform
-        
-        # run the app
-        if system == "linux2":
-            cmd = 'xdg-open "%s"' % path
+        self.log_debug("Detected system: %s" % system)
+
+        # build the commands for opening the folder on the various OS's
+        if system.startswith("linux"):
```
Comment (jfboismenu) [inline app.py]: Maybe you and @manneohrstrom should talk. You've both written the same code that opens a file. I think this should get refactored into core. Then we'll be able to reuse it in the browser integration as well.

File: app.py [hunk 34]
```diff
@@ -31,46 +32,168 @@ def init_app(self):
...
+
+        if not os.path.isfile(path):
+            raise ValueError(
+                "The path \"%s\" is not a valid file path." % path)
+
+        # get the setting
+        system = sys.platform
+        self.log_debug("Detected system: %s" % system)
+
+        if system.startswith("linux"):
```
Comment (jfboismenu) [inline app.py]: If we introduce an `open_file` function in tk-core, maybe `select_file_in_browser` wouldn't be such a bad thing either.

**Review by jfboismenu — APPROVED** [review]
Looks good, but I think you and @manneohrstrom should discuss the open file logic. It should really go inside core since we've copied that logic in at least 4 places (core, browser integration, launch app and new tk-multi-devs app)

File: app.py (hunk 33 above)
Comment (pscadding) [reply app.py]: Cheer JF, I'll bring it up with Manne, but as this code was already present (I've just altered it slightly) I'd like to get a release out now and log a ticket to refactor.

Comment (manneohrstrom) [reply app.py]: yes, sounds good @jfboismenu !

Comment (pscadding) [reply app.py]: shall I raise a new ticket for that?

Comment (manneohrstrom) [reply app.py]: i can do it as part of my [dev day ticket](https://github.com/shotgunsoftware/tk-core/pull/656) and i'll ping you when it's done!

**Comment by pscadding** [comment]
@jfboismenu I've made changes so that it uses the new core utils. I've updated the info.yml so that it requires the latest core version.
Hopefully this should be all good now?
